File: Flutter/lab2/lib/entities/user_entity.dart
class UserEntity {
  late String name;
  late String surname;
  late String dateOfBirth;
  late String address;
  late String email;
  late String male;
  late int age;
  late String phoneNumber;
  late String favCity;
  late String favCountry;
  late List<String>? favorites;

  UserEntity(
      this.name,
      this.surname,
      this.dateOfBirth,
      this.address,
      this.email,
      this.male,
      this.age,
      this.phoneNumber,
      this.favCity,
      this.favCountry);

   UserEntity.fromEmail(this.email) {
    name = "";
    surname = "";
    dateOfBirth = "";
    address = "";
    male = "";
    age = 0;
    phoneNumber = "";
    favCity = "";
    favCountry = "";
    favorites = [];
  }

  Map<String, dynamic> toJson() => {
    "name": name,
    "surname": surname,
    "dateOfBirth": dateOfBirth,
    "address": address,
    "email": email,
    "male": male,
    "age": age,
    "phoneNumber": phoneNumber,
    "favCity": favCity,
    "favCountry": favCountry,
    "favorites": favorites ?? [], // Используйте оператор ?? для предотвращения ошибок, если favorites равно null
  };

  static UserEntity fromJson(Map<String, dynamic> data) {
    return UserEntity(
        data['name'],
        data['surname'],
        data['dateOfBirth'],
        data['address'],
        data['email'],
        data['male'],
        data['age'],
        data['phoneNumber'],
        data['favCity'],
        data['favCountry']);
  }
}

File: Swift/lab3/lab3/Entities/UserEntity.swift
//
//  UserEntity.swift
//  lab3
//
//  Created by dumonten on 15/3/24.
//

import Foundation

struct UserEntity: Codable {
    var name: String
    var age: Int
    var email: String
    var surname: String
    var address: String
    var phoneNumber: String
    var dateOfBirthday: String
    var male: String
    var favCountry: String
    var favCity: String
    var favorites: [String]
    
    // Конструктор, принимающий все поля
    init(name: String, age: Int, email: String, surname: String, address: String, phoneNumber: String, dateOfBirthday: String, male: String, favCountry: String, favCity: String) {
        self.name = name
        self.age = age
        self.email = email
        self.surname = surname
        self.address = address
        self.phoneNumber = phoneNumber
        self.dateOfBirthday = dateOfBirthday
        self.male = male
        self.favCountry = favCountry
        self.favCity = favCity
        self.favorites = []
    }
    
    // Конструктор, принимающий только email
    init(email: String) {
        self.name = ""
        self.age = 0
        self.email = email
        self.surname = ""
        self.address = ""
        self.phoneNumber = ""
        self.dateOfBirthday = ""
        self.male = ""
        self.favCountry = ""
        self.favCity = ""
        self.favorites = []
    }
    
    func toJson() -> [String: Any] {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        if let jsonData = try? encoder.encode(self),
           let json = try? JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] {
            return json
        }
        return [:]
    }
    
    static func fromJson(json: [String: Any]) -> UserEntity? {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: json, options: [])
            let user = try decoder.decode(UserEntity.self, from: jsonData)
            return user
        } catch let error {
            print("Error decoding JSON: \(error)")
            return nil
        }
    }

}

